### PR TITLE
feat(unstable): Support --watch flag for bundle and fmt subcommands

### LINKED ISSUE #8181: bug(watcher): changes on a file newly added to dependency are not watched
### version

current latest master (bfce376c)

### How to reproduce

1. create a file named `main.js`with the following content:

```js
console.log('hello');
```
2. start running deno watcher by `cargo run -- run --watch --unstable main.js`. At this point we can see an output like:

```
hello
Watcher Process terminated! Restarting on file change...
```
3. create a new file named `new.js`. Nothing happens on the terminal.

```js
export const world = 'world';
```

4. edit `main.js` to the following and save it:

```diff
+import { world } from "./new.js";
console.log("hello");
+console.log(world);
```

When saving it, we get:

```
Watcher File change detected! Restarting!
hello
world
Watcher Process terminated! Restarting on file change...
```

5. edit the value of `world` in `new.js` and save it:

```diff
-export const world = "world";
+export const world = "Deno!";
```

When saving it, nothing happens although we expect `hello\nDeno!` to be printed.

I've already found the cause of this problem in the watcher implementation. This is because module resolution is done only once at the starting of the watcher. Basically we should resolve module every time any event is detected.

I'll look into this problem deeply and try to elabolate a solution.

### PR DESCRIPTION
**Pull request by magurotuna**
This PR adds `--watch` flag support for both `bundle` and `fmt` subcommands.
In addition to this, it refactors `run --watch` command so that module resolution will occur every time the file watcher detects file addition/deletion, which allows the watcher to observe a file that is newly added to the dependency as well.

-----------------------------

Fixes #8181 

Currently, `--watch` flag is working with `run` only, but it is considered the flag is going to be added to other subcommands like `fmt`, `lint`, and so on. (see #7465)
For this purpose, the current implementation of `watch_func` function has abstract, extendable design - all we have to do to add the watcher feature to subcommands is passing both target file paths and an operation you want to execute to the `watch_func` function.

Considering it deeply, it seems like this `watch_func` design is not suitable for usecases, especially `run` subcommand. In case we want to detect file changes and then restart it, the watcher should also keep an eye on deletion and addition of files or dependency.
So target file paths will vary, although the current `watch_func` receives constant file paths. This is why the issue described in #8181 happened.

To address the issue, I think we should resolve dependency whenever the watcher detects changes. Actually this is relatively heavy task to be done every time, but it seems needed.

In this PR, I've added new watcher funtcion named `watch_func_for_run` which is dedicated to `run` subcommand. 
Additionally I've removed the original `watch_func` function. I know this removal will have an effect on adding `--watch` flag to other subcommands, which are being worked on at #7740 and #7767. I think other subcommands, like `fmt`, also have the same issue as `run`. That said, if we would run `deno fmt --watch`, the watcher would have to detect any addition or deletion of files. Therefore we should probably consider again what `watch_func` function interface should look like, without using the current one.

#### TODO

- [x] add test

### CODE REVIEW

**Comment by magurotuna** [comment]
I also added `fmt --watch` feature to show how we can make subcommands other than `run` work with `--watch` flag. (a lot of code come from #7767)

**Comment by magurotuna** [comment]
Just added `bundle --watch` as well. Bundle is similar to `run` rather than `fmt`, so I renamed `watch_func_for_run` to `watch_func_with_module_resolution` and fixed its parameter type a little.

**Comment by bartlomieju** [comment]
`watch_func_with_module_resolution` - how about `watch_func_with_module_graph`?

**Comment by magurotuna** [comment]
@bartlomieju 
`bundle` operation requires `module_graph::Graph`, where `run` requires `deno_core::ModuleSpecifier`. This is why I avoided using the word `graph`as function name.  
But indeed `watch_func_with_module_resolution` is long and kind of redundant, so I'd be happy to find appropriate name...

File: cli/file_watcher.rs
Comment (bartlomieju) [inline cli/file_watcher.rs:56]: can you `Pin` the `Debounce.delay` instead?

File: cli/file_watcher.rs
Comment (bartlomieju) [inline cli/file_watcher.rs:78]: Those arguments could use some comments, it's not immediately obvious what's their purpose

File: cli/file_watcher.rs
Comment (magurotuna) [reply cli/file_watcher.rs]: I tried but I didn't know how I could make that compile, so instead I use `poll_unpin` here. How about this?

Comment (bartlomieju) [reply cli/file_watcher.rs]: Even better

File: cli/file_watcher.rs
Comment (magurotuna) [reply cli/file_watcher.rs:92]: I added some comments to describe watcher functions including their parameters. PTAL 8feed3c

**Comment by magurotuna** [comment]
Not sure the cause of flakiness in CI on macOS (pointed in #8316), but I attempted to increase timeout value.

**Review by bartlomieju — COMMENTED** [review]
@magurotuna looks good! I have a few nitpicks, but I will hold them off for now. Let's land https://github.com/denoland/deno/pull/8375 first - it should cleanup up a few bits of this PR already.

**Comment by bartlomieju** [comment]
#8375 has landed, @magurotuna could you rebase this PR?

**Comment by magurotuna** [comment]
@bartlomieju Finished rebasing. Should the PR title be changed? Maybe "feat(bundle, fmt): Support --watch flag" or something, though it would lack the information the current title has :/

File: cli/file_watcher.rs
Comment (bartlomieju) [inline cli/file_watcher.rs:26]: Let's address this TODO, how about `FileWatcherFuture`?

File: cli/file_watcher.rs
Comment (bartlomieju) [inline cli/file_watcher.rs:77]: remove `stuff`

File: cli/file_watcher.rs
Comment (bartlomieju) [inline cli/file_watcher.rs:132]: remove `stuff`

File: cli/flags.rs
Comment (bartlomieju) [inline cli/flags.rs:363]: Would be great to have unit test added for respective commands

File: cli/tests/integration_tests.rs
Comment (bartlomieju) [inline cli/tests/integration_tests.rs:1241]: Nice tests

File: cli/tests/integration_tests.rs
Comment (bartlomieju) [inline cli/tests/integration_tests.rs:1341]: 👍  great to this fixed

File: cli/main.rs
Comment (bartlomieju) [inline cli/main.rs:662]: This problem should be fixed by https://github.com/denoland/deno/pull/8328, we must be careful on the timing between the two PRs

**Review by bartlomieju — COMMENTED** [review]
@magurotuna great work! I really like the API you came up with - it prevents work duplication. I have only minor comments. I will wait on @kitsonk approval as it touches module graph.

File: cli/flags.rs
Comment (magurotuna) [reply cli/flags.rs:363]: I forgot to do that, thanks! I've added tests for `bundle --watch` and `fmt --watch`.

File: cli/file_watcher.rs
Comment (magurotuna) [reply cli/file_watcher.rs]: sounds good 👍

File: cli/file_watcher.rs
Comment (magurotuna) [inline cli/file_watcher.rs:179]: To address #7848, I fixed `watch_func_with_module_resolution` implementation so that the watcher will try to continue watching files using _previously resolved_ data.

This works fine if the target files initially have correct syntax and then updated into invalid syntax.
This also means that the watcher won't work at all if the target files initially have invalid syntax. I think it cannot be helped because the watcher cannot decide which files should be watched in this case.

Comment (magurotuna) [reply cli/file_watcher.rs:179]: If there's no problem with my approach, I'll add some integration tests to confirm this behavior. WDYT? @bartlomieju

Comment (bartlomieju) [reply cli/file_watcher.rs]: @magurotuna that sounds good! Test would be great

**Comment by magurotuna** [comment]
I added tests for `run` and `bundle` in the following cases:

- watcher exits immediately if module resolution fails at the first attempt due to invalid syntax
- watcher tries to continue watching files using previously resolved data

**Comment by magurotuna** [comment]
hmmm..... tests are sooooo flaky :(

**Comment by bartlomieju** [comment]
> hmmm..... tests are sooooo flaky :(

@magurotuna are you talking about tests for this PR or tests in general?

**Comment by magurotuna** [comment]
@bartlomieju 
Seems like both of them.

1. https://github.com/denoland/deno/runs/1413182650#step:16:3855 -> operation timeout on macOS, which happened at the place unrelated to this PR
2. https://github.com/denoland/deno/runs/1416308269#step:14:12158 -> rusty_v8 build failed on ubuntu, totally unrelated to this PR
3. https://github.com/denoland/deno/runs/1416415130#step:16:3862 -> `fmt_watch_test` failed on macOS. This seems to be related to this PR and #8316.

I have no idea why `fmt_watch_test` failed. I tried to run the tests on my local mac, they were totally fine.

**Comment by bartlomieju** [comment]
@magurotuna thanks for detailed description. I'm aware of the first problem, but haven't found the reason yet. Second failure seems like an intermittent failure. As for third I'm not sure, let's keep a close eye on it.

**Comment by magurotuna** [comment]
Several times of resolving conflict have been a little bit grunt work 😢

**Comment by bartlomieju** [comment]
> Several times of resolving conflict have been a little bit grunt work 😢

Sorry for that @magurotuna, I'll try to land your PR this weekend.

Can I ask you to take a look [at this comment](https://github.com/denoland/deno/pull/8276#discussion_r523464871) and check if that's still the case?

**Comment by magurotuna** [comment]
@bartlomieju Sure, I'll check for that.

**Comment by bartlomieju** [comment]
@magurotuna I've tested this PR locally and it's working very well! Code in this PR is a big improvement compared to current implementation and LGTM. Please prepare commit message before landing :)

File: cli/main.rs
Comment (magurotuna) [reply cli/main.rs]: I tried to check if this is still the case or not, but it seemed impossible to create `ProgramState` only once before the watcher starts, because of lifetime and trait bound restriction.
So effectively `ProgramState` ended up being created every time the watcher restarts. I think it's okay, so just removed the FIXME comment.

**Comment by magurotuna** [comment]
@bartlomieju 
Thanks for your test!

> Please prepare commit message before landing :)

Sorry if I misunderstand it, but does it mean that you want me to write a commit message that will be used in a squashed merge commit?

**Comment by bartlomieju** [comment]
> Sorry if I misunderstand it, but does it mean that you want me to write a commit message that will be used in a squashed merge commit?

Yes, exactly

**Comment by magurotuna** [comment]
@bartlomieju 
Sure, I've added a brief message to the head of the PR's body. Is this fine?

**Review by bartlomieju — APPROVED** [review]
LGTM

**Pull request merged**
